mempool/mempool.go

```diff
@@ -337,6 +349,7 @@ func (mem *Mempool) CheckTx(tx types.Tx, cb func(*abci.Response)) (err error) {
 	if err = mem.proxyAppConn.Error(); err != nil {
 		return err
 	}
+	mem.proxyAppConn.SetResponseCallback(mem.resCb(txInfo.PeerID))
```

Comment:
so this works for a the `localClient` where `Async` calls are actually synchronous, but for the `socketClient`, isn't it possible that this could overwrite the response callback for a call that hasn't returned yet? ie. we call `CheckTxWithInfo(tx1, cb, txInfo1)` and then call `CheckTxWithInfo(tx2, cb, txInfo2)` before the response comes back from tx1 - then the callback will have been overwritten and when resCb executes it will use `txInfo2` for processing the response of both `tx1` and `tx2`! Unless I'm missing something, we might need to rethink the approach here...

Reply:
My understanding was that every `CheckTxAsync` creates a new ReqRes with the currently known callback. So that ReqRes object knows its own callback, there isn't a global per-connection callback. Looking at socket_client, I have no idea where the callback gets set on that created ReqRes, so your probably right. https://github.com/tendermint/tendermint/blob/dev/mempool_no_gossip_back/abci/client/socket_client.go#L246

Reply:
there's two callbacks, a global one (set on the proxy app) and a per-call one (set on the reqres). this is setting the global one, but maybe we need to fix the mempool architecture to set it on the reqres

Reply:
Makes sense, though I still find the way that global one is handled kind of odd. This shouldn't be hard regardless. We already set the reqres callback if a callback is supplied. I can just always set on request and compose the two if a callback is supplied.

Reply:
resolved in latest commit